### PR TITLE
refactor(mui-datatables): updated for v3.1.x

### PR DESCRIPTION
**Pull request by favna**
1. Noticed a missing key on `textLabels.body`, this PR adds it
2. Updated the package to support `mui-datatables` version 3.1.x
3. Also covered the changes to resolve #45264 - no idea what OP there meant with "when I find more cycles", but after @danger-public alerted me of the missing property I figured I should add it.

﻿Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/gregnb/mui-datatables/#localization & https://github.com/gregnb/mui-datatables/blob/master/docs/v2_to_v3_guide.md
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)

### CODE REVIEW

**Comment by Favna** [comment]
Drafting this so I can include the changes made in mui-datatables v3.0 before it gets merged. That way it's all inclusive.

**Comment by Favna** [comment]
Undrafted, PR now covers the changes for version 3.0.0 as well.

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## mui-datatables ([<kbd>unpkg</kbd>](https://unpkg.com/browse/mui-datatables@latest/))
was missing the following properties:
1. DebounceTableSearch
1. debounceSearchRender
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 467e20d16b84479eee23af85a0228f002fd2d04c
</p>

**Comment by Favna** [comment]
Since this PR was drafted then undrafted it seems like it got reset for the 7 day turn around cycle, seeing its status on the project board. Just going to ping the other maintainers in here like @typescript-bot does for review so this can get merged:

🔔   @ankithkonda @hwatersiv @souppower @byrekt - please [review this PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45306/files) in the next few days. Be sure to explicitly select **`Approve`** or **`Request Changes`** in the GitHub UI so I know what's going on.

**Comment by elibarzilay** [comment]
Reping @ankithkonda @hwatersiv @souppower @byrekt

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:146]: can you left this as is and mark wiht proper JSDOC? What is better for users?
```
/** @deprecated */
```
this will allow users to recognize the fact that code changed:
https://github.com/gregnb/mui-datatables/blob/49c18a4ccffeccc684c6091cae968af441d2568b/src/MUIDataTable.js#L510

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:204]: same as with sort order, should that one be only marked wiht `@deprecated` JSDoc?
https://github.com/gregnb/mui-datatables/blob/3.1.4/docs/v2_to_v3_guide.md#fixedheaderoptions-is-now-fixedheader-and-fixedselectcolumn
> The table will still accept the fixedHeaderOptions option, but it will print a deprecation notice when run in development mode.

Both ways of handling this are acceptable I think, but usually deprecation allows to migrate code in somehow slower fashion compared to tsc approach.

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:191]: same topic of `@deprecated`:
https://github.com/gregnb/mui-datatables/blob/3.1.4/docs/v2_to_v3_guide.md#disabletoolbarselect-deprecated-in-favor-of-selecttoolbarplacement
> disableToolbarSelect will still work, but will give a deprecation warning

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:217]: same `@deprecated`
> The old onColumnViewChange option will still work but will give a deprecation warning.

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:225]: this should be optional IMO:
https://github.com/gregnb/mui-datatables/blob/3ae8302d4c006ee47f951ac0ba113b7fb1b0eba2/src/MUIDataTable.js#L1170
```suggestion
    onColumnSortChange?: (changedColumn: string, direction: 'asc' | 'desc') => void;
```

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:224]: should stay optional IMO:
https://github.com/gregnb/mui-datatables/blob/3ae8302d4c006ee47f951ac0ba113b7fb1b0eba2/src/MUIDataTable.js#L1188
```suggestion
    onChangeRowsPerPage?: (numberOfRows: number) => void;
```

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:223]: should stay optional IMO:
https://github.com/gregnb/mui-datatables/blob/3ae8302d4c006ee47f951ac0ba113b7fb1b0eba2/src/MUIDataTable.js#L1202
```suggestion
    onChangePage?: (currentPage: number) => void;
```

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:231]: should stay optional IMO:
https://github.com/gregnb/mui-datatables/blob/ed3b8e38889d061a8cc858637b1d1dfe0fa55556/src/utils.js#L99
```suggestion
    onDownload?: (
```

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:237]: shuld stay optional IMO:
https://github.com/gregnb/mui-datatables/blob/3ae8302d4c006ee47f951ac0ba113b7fb1b0eba2/src/MUIDataTable.js#L1258
https://github.com/gregnb/mui-datatables/blob/3ae8302d4c006ee47f951ac0ba113b7fb1b0eba2/src/MUIDataTable.js#L195
```suggestion
    onFilterChange?: (changedColumn: string, filterList: any[], type: FilterType | 'chip' | 'reset') => void;
```

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:1]:
```suggestion
// Type definitions for mui-datatables 3.1
```
Let's use the latest one for review, sounds OK?

File: types/mui-datatables/index.d.ts
Comment (Favna) [reply types/mui-datatables/index.d.ts:225]: The entire type is wrapped in `Partial<>`

**Review by peterblazejewicz — CHANGES_REQUESTED** [review]
@Favna Jeroen, can you review those comments about `@deprecated` tag (optional, up to you how to approach the problem here, but I would probably follow the native JS way) and about those optional props changes? (there are more then I've marked).
When done ping me, I've read a thread in other place about this PR, thanks!

**Comment by peterblazejewicz** [comment]
just FYI, why `@deprecate` is important (and better supported):
https://devblogs.microsoft.com/typescript/announcing-typescript-4-0-beta/#deprecated-support

**Comment by Favna** [comment]
@peterblazejewicz deprecation JSDoc added and changed version in 1e3fa01fca62c1e54b1467d43d753a146b5e65e0. As for the optional stuff, the entire type is wrapped in `Partial<>` hence why I removed all the `?:`'s.

**Comment by peterblazejewicz** [comment]
@Favna sorry for not noticing the `Partial` (there is no 🤦  reaction LoL), give me some time, will revisit

**Comment by Favna** [comment]
If you give me 10'ish minutes, I see some more changes were made to the lib between 3.0.0 and 3.1.x. I'll see into adding those as well.

**Comment by Favna** [comment]
@peterblazejewicz Okay ready for review now

File: types/mui-datatables/index.d.ts
Comment (peterblazejewicz) [inline types/mui-datatables/index.d.ts:547]: isn't the `debounceWait` optional (defaults to 200ms)?
https://github.com/gregnb/mui-datatables/blob/468c9c7e339501a786caf7b6afbea2c4f1f63a6f/src/plug-ins/DebounceSearchRender.js#L103
```suggestion
export function debounceSearchRender(debounceWait?: number): MUIDataTableOptions['customSearchRender'];
```

**Comment by Favna** [comment]
@peterblazejewicz I'm not sure what the goal is of the exported function (I've not used the plugin myself) but I assume the value is actually read at https://github.com/gregnb/mui-datatables/blob/468c9c7e339501a786caf7b6afbea2c4f1f63a6f/src/plug-ins/DebounceSearchRender.js#L67-L71

File: types/mui-datatables/index.d.ts
Comment (Favna) [reply types/mui-datatables/index.d.ts]: yeah good catch

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## uikit ([<kbd>unpkg</kbd>](https://unpkg.com/browse/uikit@latest/))
was missing the following properties:
1. The declaration doesn't match the JavaScript module 'uikit'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.

The most common way to resolve this error is to use 'export =' syntax.
To learn more about 'export =' syntax, see https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require.
1. util
1. data
1. prefix
1. options

<details>
<summary>as well as these 1 other properties...</summary>
<p>version</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against bf4d1163f56db34f7d4afd80ac6b493c35a4cb0f
</p>

**Comment by Favna** [comment]
uuuhhh... looks like @danger-public got the wrong pull request or something lol.

**Comment by peterblazejewicz** [comment]
@Favna  
[optional]
if you rebase your branch on master, it should cleanly build, now the diff checker detects outer changes. Tested locally:
```git rebase master
First, rewinding head to replay your work on top of it...
Applying: refactor(mui-datatables): added extra key to textLabels
Applying: refactor(mui-datatables): update mui-datatables to v3.0.0
Applying: refactor(mui-datatables): include debounce search render plugin
Applying: address peter's review
Applying: add missing types since 3.0.x
```
(so it rebases cleanly)
and diffs for tests:
```
Running: git diff master --name-status
M	types/mui-datatables/index.d.ts
M	types/mui-datatables/mui-datatables-tests.tsx
M	types/mui-datatables/tsconfig.json
Testing 1 changed packages: mui-datatables
Testing 0 dependent packages: 
Installing NPM dependencies...
```

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
I've tried samples from v3.* README

@Favna thx for the update!

**Comment by Favna** [comment]
@peterblazejewicz rebased on master and squashed

**Review by peterblazejewicz — APPROVED** [review]

**Comment by occult** [comment]
Thank you so much! It will help out a lot. LGTM :)

**Comment by AntonioLobo** [comment]
LGTM!

**Comment by peterblazejewicz** [comment]
👯

**Comment by peterblazejewicz** [comment]
fixed #45306

**Pull request merged**
